src/sql/workbench/contrib/notebook/test/browser/notebookEditor.test.ts

```diff
@@ -422,11 +376,358 @@ suite('Test class NotebookEditor', () => {
+		};
+		notebookEditor['_input'] = <NotebookInput>{};
+		notebookEditor['_updateFinderMatchState']();
+		notebookEditor['_input'] = undefined;
+		notebookEditor['_updateFinderMatchState']();
+	});
+
+	test(`Verifies onFindCountChange.fire invokes _updateFinderMatchState`, async () => {
+		await setupNotebookEditor(notebookEditor, untitledNotebookInput);
+		untitledNotebookInput.notebookFindModel.notebookModel = undefined; // clear preexisting notebookModel
+		await notebookEditor.setNotebookModel();
+		const findState = notebookEditor['_findState'];
+		const newState: INewFindReplaceState = {
+			searchString: getRandomString(1, 10),
+			matchCase: true,
+			wholeWord: true,
+			searchScope: <NotebookRange>{}
+		};
+		findState.change(newState, false); //installs _updateFinderMatchState as event handler for onFindCountChange event
+		let updateFinderMatchStateCalled = false;
+		notebookEditor['_updateFinderMatchState'] = () => {
+			updateFinderMatchStateCalled = true;
+		};
+		notebookEditor.notebookInput.notebookFindModel['_onFindCountChange'].fire(null);
+		assert.strictEqual(updateFinderMatchStateCalled, true, `_updateFinderMatchState() should have been called`);
 	});
 });
 
+async function findStateChangeSetup(instantiationService: TestInstantiationService, workbenchThemeService: any, notebookService: NotebookService, untitledNotebookInput: UntitledNotebookInput, modelFindExpression: string, currentMatch: NotebookRange, searchString: string, wholeWord: boolean, matchCase: boolean, searchScope: NotebookRange | null = undefined, findMatches: Array<NotebookFindMatch> = []) {
+	const findReplaceStateChangedEvent: FindReplaceStateChangedEvent = {
+		searchString: searchString !== undefined,
+		matchCase: matchCase,
+		wholeWord: wholeWord,
+		searchScope: !!searchScope,
+		isRevealed: true,
+
+		moveCursor: false,
+		updateHistory: false,
+		replaceString: false,
+		isReplaceRevealed: false,
+		isRegex: false,
+		preserveCase: false,
+		matchesPosition: false,
+		matchesCount: false,
+		currentMatch: false,
+		loop: false
+	};
+
+	const notebookEditor = createNotebookEditor(instantiationService, workbenchThemeService, notebookService);
+	await setupNotebookEditor(notebookEditor, untitledNotebookInput);
+	untitledNotebookInput.notebookFindModel.notebookModel = undefined; // clear preexisting notebookModel
+	await notebookEditor.setNotebookModel();
+	const findState = notebookEditor['_findState'];
+	findState['_searchString'] = searchString;
+	findState['_isRevealed'] = findReplaceStateChangedEvent.isRevealed;
+	findState['_wholeWord'] = findReplaceStateChangedEvent.wholeWord;
+	findState['_matchCase'] = findReplaceStateChangedEvent.matchCase;
+	findState['_searchScope'] = searchScope;
+	const notebookFindModel = notebookEditor.notebookInput.notebookFindModel;
+	notebookFindModel['_findMatches'] = findMatches;
+	notebookFindModel['_findArray'] = findMatches.map(x => x?.range);
```

Comment:
You have a ton of direct property accessors in these tests - generally we should strive to do that as little as possible. Normally we accomplish this by having Test/Stub classes that extend the base class and use protected fields to modify these as needed.

Reply:
but I need these for testing private methods as well which cannot be accessed from an extended class.

Reply:
You shouldn't "need" to test private methods. Private methods should generally be implementation details, what you should be testing is the public API and then verifying that the state of the application is what you expect afterwards.

Now granted there are exceptions since we haven't been the greatest about setting stuff up to be testable (such as setting up dependency injection fully) - but when I see 10+ calls to private members of a class it makes me think that something could be improved from the test side.

I'll take a look at this in more detail later and see if I can see any examples of places with room for improvements - but don't block on this for now.

Reply:
So just to make sure. To implement this suggestion, I will be changing a whole bunch of 'private' members in class NotebookEditor (the class under test) to 'protected'. So that in an extended stub class, I can provide accessors to these members.

Reply:
I think it's worth investigating - but if it ends up being a ton of extra work then probably not worth the effort.

This to me is more of an indication that either the tests are testing the wrong thing or the code isn't architected well. As I mentioned in an ideal world we shouldn't need to be testing/modifying the internal state of the things we're testing. The fact that you need to do so - and in so many cases - tells me that improvements could be made here in that regard and I would suggest focusing on that if you're going to be writing tests. Remember that we want it to be easy to add more tests later on too - and that gets a lot harder when we start relying on internal state changes to do what we want.

Reply:
At this stage it will be a ton of rework. In the interest of spending the developement dollars on wider coverage I will move to test other areas. We can clean these up in future. But I will incorporate this valuable feedback in future tests.